### PR TITLE
Bug 1691660: pkg/controller: always use the OSImageURL from the CVO

### LINKED ISSUE #465: Prevent OSImageURL from being set by users
I'd love a way to prevent people from creating machineconfigs that set osimageurl (and/or edit the current ones?). Ideally, I can see the CVO driving that w/o user input at all. This will reduce the chance for users to do something bad (in my view, if it's not wrong).
We could detect if a machineconfig has an osimageurl and just override it with what we have in the controllerconfig resulting in a no-op for users.

### PR DESCRIPTION
**Pull request by runcom**
Signed-off-by: Antonio Murdaca <runcom@linux.com>

**- What I did**

Always use the osimageurl coming from the CVO. 

close #465 

**- How to verify it**

Added a unit test to cover this change 

**- Description for the changelog**

### CODE REVIEW

File: pkg/controller/render/render_controller.go [hunk 1]
```diff
@@ -434,13 +440,20 @@ func (ctrl *Controller) garbageCollectRenderedConfigs(pool *mcfgv1.MachineConfig
...
-	generated, err := generateMachineConfig(pool, configs)
+	cc, err := ctrl.ccLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("could not enumerate ControllerConfig %s", err)
+	}
+	if len(cc) == 0 {
+		return fmt.Errorf("ControllerConfigList is empty")
+	}
+
+	generated, err := generateMachineConfig(pool, configs, cc[0])
```
Comment (runcom) [inline pkg/controller/render/render_controller.go]: maybe someone can shed some light on this snipped I found in the kubeconfig/containerruntimeconfig controllers. Is it correct to grab the controllerconfig like this? isn't there always just 1 controller config at any point in time?

Comment (runcom) [reply pkg/controller/render/render_controller.go]: @cgwalters @kikisdeliveryservice ^^^

**Comment by kikisdeliveryservice** [comment]
Taking a look at this, there've been a couple PRs that change the render/controller config functions so refamiliarizing myself.

File: pkg/controller/render/render_controller.go (hunk 1 above)
Comment (cgwalters) [reply pkg/controller/render/render_controller.go]: Yeah, AFAIK there's always just one.  It's basically a way for the template controller to react to just one aggregation of data AIUI.

But I don't see the problem with having the render controller grab it directly like this either.

**Comment by cgwalters** [comment]
/approve
I like the idea though am uncertain about landing this right now given the testing we've done with the existing implementation.  But it looks generally good to me.

**Comment by runcom** [comment]
I'm uncertain about landing this as well, but w/o this (assuming it works and I can add e2e) customers and users will be able to upgrade/downgrade the baseos through MCs which id love to avoid

**Comment by kikisdeliveryservice** [comment]
q: are we _sure_ we don't want to allow this at all? just wondering before we go to the trouble of landing this and then having it reversed..

that said, i think this should be held until later when the os upgrades are healthy generally.

**Comment by runcom** [comment]
> q: are we _sure_ we don't want to allow this at all?

I don't see any real value in allowing them doing something like that, we want to control upgrades (generally) through CVO, i've always seen osimageurl in the MCs like something for testing for now

**Comment by runcom** [comment]
This has become more relevant in some hours as upgrades should follow the pattern where master are always upgraded before workers...we can play and check what users pass us of course and compare between pools but I feel the cleanest way to do it is to just forbid osimageurl altogether

**Comment by runcom** [comment]
question now is, since #480, do we want to go out with the possibility for users to do this https://github.com/openshift/machine-config-operator/issues/183#issuecomment-463668093 and then break them?

File: pkg/apis/machineconfiguration.openshift.io/v1/helpers.go [hunk 2]
```diff
@@ -12,25 +13,24 @@ import (
...
 
-	outOSImageURL := configs[0].Spec.OSImageURL
 	outIgn := configs[0].Spec.Config
 	for idx := 1; idx < len(configs); idx++ {
-		config := configs[idx]
-		if outOSImageURL == "" {
-			outOSImageURL = config.Spec.OSImageURL
+		mc := configs[idx]
+		if mc.Spec.OSImageURL != osImageURL {
+			glog.Infof("Detected OSImageURL %q in MachineConfig %q which is not what provided by the CVO %q", mc.Spec.OSImageURL, mc.Name, osImageURL)
```
Comment (LorbusChris) [inline pkg/apis/machineconfiguration.openshift.io/v1/helpers.go]: nit: what *is* provided

**Comment by runcom** [comment]
removing WIP here, up for further discussion as I do really feel we shouldn't really expose this

**Comment by runcom** [comment]
The development flow that this might break could be handled through #496 overriding `machine-os-content` directly (even if you can't really control the pool, but hey, we want https://github.com/openshift/machine-config-operator/issues/480 anyway)

**Comment by cgwalters** [comment]
I think my vote on this is to land tests for OS updates first.  The code here looks reasonable but it's also a nontrivial change; we know today that `oc adm upgrade` works.  The problem is that doing testing here nicely depends on https://github.com/openshift/machine-config-operator/pull/496

**Comment by cgwalters** [comment]
> removing WIP here, up for further discussion as I do really feel we shouldn't really expose this

I understand that but on the other hand...an admin really has to jump through hoops today to actually override `osImageURL`.

**Comment by runcom** [comment]
> I think my vote on this is to land tests for OS updates first.

I'm super ok with this of course

**Comment by cgwalters** [comment]
One thing I think we could envision now potentially is adding a new `custommachineconfig` CRD or so that is like `machineconfig` but for users...and wouldn't have `osImageURL`.

Or alternatively, move our rendered ones into a new `renderedmachineconfig` CRD?

**Comment by runcom** [comment]
The second option sounds very reasonable to me :) (having a rendered MC CRD)

**Comment by runcom** [comment]
Related to https://bugzilla.redhat.com/show_bug.cgi?id=1691660 now.
Not sure how many users are now using osimageurl directly in machineconfigs, now that upgrades are generally working fine. 
The related Bugzilla just highlights a flaw where if we have an user provided MC with osimageurl, then we don't respect the machineoscontent from the payload, which is wrong. I don't think there's a nice way to still honor the payload while leaving an osimageurl MC around so probably the time to just get rid of it (or pursue the ideas Colin had in previous comments) 

Any ideas?

**Comment by runcom** [comment]
This is rebased now and ready to go!

**Comment by runcom** [comment]
We could also just get rid of OSImageURL all together in machineconfigs to avoid any confusion. Haven't checked but we just get the OSImageURL from the controller config.

**Pull request merged**